### PR TITLE
Make local_directory a required argument for spilling impls

### LINKED ISSUE #1018: Program crashes when running same program twice sequentially
When running the below piece of code:

```
import cupy as cp
import numpy as np
import dask.array as da
from dask_cuda import LocalCUDACluster
from dask.distributed import Client, wait
import rmm

if __name__ == '__main__':

    cluster = LocalCUDACluster('0', rmm_managed_memory=True)
    client = Client(cluster)
    client.run(cp.cuda.set_allocator, rmm.rmm_cupy_allocator)

    # Here we set RMM/CuPy memory allocator on the "current" process,
    # i.e., the Dask client.
    rmm.reinitialize(managed_memory=True)
    cp.cuda.set_allocator(rmm.rmm_cupy_allocator)

    shape = (512, 512, 30000)
    chunks = (100, 100, 1000)

    huge_array_gpu = da.ones_like(cp.array(()), shape=shape, chunks=chunks)
    array_sum = da.multiply(huge_array_gpu, 17).persist()
    # `persist()` only does lazy evaluation, so we must `wait()` for the
    # actual compute to occur.
    wait(array_sum)

```
It runs perfectly the first time around. This of course creates a  folder called ```dask-worker-space/storage```. If I delete this folder, I can run the program again with no problem. If I do not, however, I get the following error:

```~/Desktop/src/pygpubatch$ python3 regex.py
2022-10-18 10:10:16,404 - distributed.preloading - INFO - Creating preload: dask_cuda.initialize
2022-10-18 10:10:16,404 - distributed.preloading - INFO - Import preload module: dask_cuda.initialize
2022-10-18 10:10:16,626 - distributed.nanny - ERROR - Failed to start worker
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.
2022-10-18 10:10:16,676 - distributed.nanny - ERROR - Failed while trying to start worker process: Worker failed to start.
2022-10-18 10:10:16,677 - distributed.nanny - ERROR - Failed to connect to process
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 669, in start
    msg = await self._wait_until_connected(uid)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 789, in _wait_until_connected
    raise msg["exception"]
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.
2022-10-18 10:10:16,678 - distributed.nanny - ERROR - Failed to start process
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 423, in instantiate
    result = await self.process.start()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 669, in start
    msg = await self._wait_until_connected(uid)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 789, in _wait_until_connected
    raise msg["exception"]
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.
Task exception was never retrieved
future: <Task finished name='Task-22' coro=<_wrap_awaitable() done, defined at /home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py:681> exception=RuntimeError('Nanny failed to start.')>
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 350, in start_unsafe
    response = await self.instantiate()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 423, in instantiate
    result = await self.process.start()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 669, in start
    msg = await self._wait_until_connected(uid)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 789, in _wait_until_connected
    raise msg["exception"]
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 688, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Nanny failed to start.
2022-10-18 10:10:16,683 - tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOLoop object at 0x7fcf80946fd0>>, <Task finished name='Task-21' coro=<SpecCluster._correct_state_internal() done, defined at /home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/deploy/spec.py:319> exception=RuntimeError('Worker failed to start.')>)
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/tornado/ioloop.py", line 765, in _discard_future_result
    future.result()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/deploy/spec.py", line 358, in _correct_state_internal
    await w  # for tornado gen.coroutine support
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 469, in start
    raise self.__startup_exc
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 350, in start_unsafe
    response = await self.instantiate()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 423, in instantiate
    result = await self.process.start()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 669, in start
    msg = await self._wait_until_connected(uid)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 789, in _wait_until_connected
    raise msg["exception"]
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.
2022-10-18 10:10:17,673 - distributed.preloading - INFO - Creating preload: dask_cuda.initialize
2022-10-18 10:10:17,673 - distributed.preloading - INFO - Import preload module: dask_cuda.initialize
2022-10-18 10:10:17,871 - distributed.nanny - ERROR - Failed to start worker
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.
2022-10-18 10:10:17,915 - distributed.nanny - ERROR - Failed while trying to start worker process: Worker failed to start.
2022-10-18 10:10:17,915 - distributed.nanny - ERROR - Failed to connect to process
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 669, in start
    msg = await self._wait_until_connected(uid)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 789, in _wait_until_connected
    raise msg["exception"]
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.
2022-10-18 10:10:17,916 - distributed.nanny - ERROR - Failed to start process
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 423, in instantiate
    result = await self.process.start()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 669, in start
    msg = await self._wait_until_connected(uid)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 789, in _wait_until_connected
    raise msg["exception"]
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.
Task exception was never retrieved
future: <Task finished name='Task-36' coro=<_wrap_awaitable() done, defined at /home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py:681> exception=RuntimeError('Nanny failed to start.')>
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 350, in start_unsafe
    response = await self.instantiate()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 423, in instantiate
    result = await self.process.start()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 669, in start
    msg = await self._wait_until_connected(uid)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 789, in _wait_until_connected
    raise msg["exception"]
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 688, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Nanny failed to start.
Traceback (most recent call last):
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1406, in start_unsafe
    await self._register_with_scheduler()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in _register_with_scheduler
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/worker.py", line 1093, in <dictcomp>
    types={k: typename(v) for k, v in self.data.items()},
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/_collections_abc.py", line 850, in __iter__
    for key in self._mapping:
RuntimeError: Set changed size during iteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/joachim/Desktop/src/pygpubatch/regex.py", line 10, in <module>
    cluster = LocalCUDACluster('0', rmm_managed_memory=True)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/dask_cuda/local_cuda_cluster.py", line 366, in __init__
    self.sync(self._correct_state)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/utils.py", line 338, in sync
    return sync(
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/utils.py", line 405, in sync
    raise exc.with_traceback(tb)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/utils.py", line 378, in f
    result = yield future
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/deploy/spec.py", line 358, in _correct_state_internal
    await w  # for tornado gen.coroutine support
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 469, in start
    raise self.__startup_exc
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 480, in start
    await asyncio.wait_for(self.start_unsafe(), timeout=timeout)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 350, in start_unsafe
    response = await self.instantiate()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 423, in instantiate
    result = await self.process.start()
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 669, in start
    msg = await self._wait_until_connected(uid)
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 789, in _wait_until_connected
    raise msg["exception"]
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/nanny.py", line 858, in run
    await worker
  File "/home/joachim/anaconda3/envs/rps/lib/python3.9/site-packages/distributed/core.py", line 488, in start
    raise RuntimeError(f"{type(self).__name__} failed to start.") from exc
RuntimeError: Worker failed to start.
```

### PR DESCRIPTION
**Pull request by wence-**
For automated cleanup when the cluster exits, the on-disk spilling directory needs to live inside the relevant worker's local_directory. Since we do not have a handle on the worker when constructing the keyword arguments to DeviceHostFile or ProxifyHostFile, instead take advantage of dask/distributed#7153 and request that we are called with the worker_local_directory as an argument. Closes #1018.

### CODE REVIEW

**Comment by wence-** [comment]
This is a breaking change (since it is no longer possible to provide a `None` local_directory to the various spilling classes and get some kind of default). On the other hand, as #1018 demonstrates, there is a need to put stuff in the appropriate worker's local directory (and that location can still be controlled by configuration via usual dask mechanisms).

**Comment by wence-** [comment]
cc @madsbk as implementor of most of these classes.

**Review by pentschev — COMMENTED** [review]
Left a few questions, but otherwise looks good. Also would be good to wait for @madsbk 's review as well, in case he notices anything that may not play well with proxied/JIT-Unspill.

**Review by madsbk — COMMENTED** [review]
LGTM. 
I know this has not been changed in this PR, but I would suggest that we use the `tmp_path` fixture instead of `tempfile.TemporaryDirectory()`.

**Comment by wence-** [comment]
> LGTM, but do we need a fallback in the case Distributed doesn't provide the `worker_local_directory` argument?

Ah, if the user is running with an older version of distributed. Yes, good point.

> Also, I know this has not been changed in this PR, but I would suggest that we use the `tmp_path` fixture instead of `tempfile.TemporaryDirectory()`.

I will have a look.

**Comment by pentschev** [comment]
> > LGTM, but do we need a fallback in the case Distributed doesn't provide the `worker_local_directory` argument?
> 
> Ah, if the user is running with an older version of distributed. Yes, good point.

I don't think we need one. We are always moving up the Distributed requirement in Dask-CUDA, i.e., we don't provide backwards compatibility.

File: dask_cuda/cuda_worker.py [hunk 1]
```diff
@@ -175,7 +175,6 @@ def del_pid_file():
                         device_memory_limit, device_index=i
                     ),
                     "memory_limit": memory_limit,
-                    "local_directory": local_directory,
```
[the remark this answers is not in the document]
Comment (wence-) [reply dask_cuda/cuda_worker.py:178]: No, I think this is one we want the user to be able to control. The point with the plugins is that each worker, when it boots makes something relative to `local_directory / dask-worker-space / worker-SOMERANDOMSTRING` and we don't know what `SOMERANDOMSTRING` is (but we do still want control over `local_directory`.

File: dask_cuda/local_cuda_cluster.py [hunk 2]
```diff
@@ -289,7 +289,6 @@ def __init__(
                     {
                         "device_memory_limit": self.device_memory_limit,
                         "memory_limit": self.memory_limit,
-                        "local_directory": local_directory,
```
[the remark this answers is not in the document]
Comment (wence-) [reply dask_cuda/local_cuda_cluster.py:292]: As above, I think this is OK.

**Comment by wence-** [comment]
> I will have a look.

Done. Via a module-scope fixture.

**Comment by pentschev** [comment]
It seems tests are failing because [Distributed nightly gets installed](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.2,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/968/consoleFull#L969), which is [posteriorly overwritten by the current release](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.2,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/968/consoleFull#L2,272). This happens because [Distributed nightly is installed in ](https://github.com/rapidsai/dask-cuda/blob/3e5a19b246aa684d2bedfb4673960d895d81dd24/ci/gpu/build.sh#L72-L76)build.sh, and I guess that [Dask-CUDA’s conda package that is later built and installed](https://github.com/rapidsai/dask-cuda/blob/branch-22.12/ci/gpu/build.sh#L93-L104) grabs Distributed stable again. I'm asking internally why this is happening this way.

**Comment by wence-** [comment]
> It seems tests are failing because [Distributed nightly gets installed](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.2,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/968/consoleFull#L969), which is [posteriorly overwritten by the current release](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.2,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/968/consoleFull#L2,272). This happens because [Distributed nightly is installed in ](https://github.com/rapidsai/dask-cuda/blob/3e5a19b246aa684d2bedfb4673960d895d81dd24/ci/gpu/build.sh#L72-L76)build.sh, and I guess that [Dask-CUDA’s conda package that is later built and installed](https://github.com/rapidsai/dask-cuda/blob/branch-22.12/ci/gpu/build.sh#L93-L104) grabs Distributed stable again. I'm asking internally why this is happening this way.

Thanks @pentschev!

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
Failing tests are unrelated, opened an issue in https://github.com/rapidsai/dask-cuda/issues/1032 .

**Comment by pentschev** [comment]
rerun tests

**Comment by wence-** [comment]
These look like actually my fault, let me fix.

**Comment by wence-** [comment]
> These look like actually my fault, let me fix.

Actually perhaps not, can't reproduce locally.

**Comment by pentschev** [comment]
@wence- just gave it a quick try and I can reproduce the failure using `docker.io/gpuci/rapidsai:22.12-cuda11.2-devel-centos7-py3.8 ` (CI test that fails) but not `docker.io/gpuci/rapidsai:22.12-cuda11.5-devel-ubuntu20.04-py3.9` (one of the CI tests that pass). The difference seems to hint at something with Python 3.8, but not sure to be honest.

**Comment by wence-** [comment]
> @wence- just gave it a quick try and I can reproduce the failure using `docker.io/gpuci/rapidsai:22.12-cuda11.2-devel-centos7-py3.8 ` (CI test that fails) but not `docker.io/gpuci/rapidsai:22.12-cuda11.5-devel-ubuntu20.04-py3.9` (one of the CI tests that pass). The difference seems to hint at something with Python 3.8, but not sure to be honest.

Thanks, I think I fixed this. It was due to use of `typing.MutableMapping` which breaks `inspect.getfullargspec` in Python 3.8. I've implemented some workarounds.

**Review by pentschev — APPROVED** [review]
This is great, thanks @wence- for figuring out the issue and for the fix. LGTM.

**Comment by pentschev** [comment]
@gpucibot merge

File: dask_cuda/device_host_file.py [hunk 3]
```diff
@@ -234,6 +235,34 @@ def __init__(
         # For Worker compatibility only, where `fast` is host memory buffer
         self.fast = self.host_buffer if memory_limit is None else self.host_buffer.fast
 
+    if sys.version_info > (3, 8):
```
[the remark this answers is not in the document]
Comment (wence-) [reply dask_cuda/device_host_file.py:238]: Because 3.8.1 > 3.8, but fixed in #1036

**Pull request merged**